File: src/main/scala/2017/day-11.scala
package adventofcode2017

import scala.io.Source

object Day11 {
  lazy val input: String = Source.fromFile(s"2017/input/day-11.input").mkString.trim

  val map = Map[String, ((Int, Int, Int, Int)) => ((Int, Int, Int, Int))](
    "n" -> { case ((n, s, e, w)) => ((n+2, s-2, e, w)) },
    "ne" -> { case ((n, s, e, w)) => ((n+1, s-1, e+1, w-1)) },
    "se" -> { case ((n, s, e, w)) => ((n-1, s+1, e+1, w-1)) },
    "s" -> { case ((n, s, e, w)) => ((n-2, s+2, e, w)) },
    "sw" -> { case ((n, s, e, w)) => ((n-1, s+1, e-1, w+1)) },
    "nw" -> { case ((n, s, e, w)) => ((n+1, s-1, e-1, w+1)) }
  )

  def getDistance(pos:(Int, Int, Int, Int)): Int = {
    pos match {
      case (n, s, e, w) => 
        val absn = Math.abs(n)
        val absw = Math.abs(w)
        absw + (absn - absw) / 2
    }
  }

  def main(args: Array[String]): Unit = {
    val steps = input.split(",").toList
    val (pos, max) = steps.foldLeft(((0, 0, 0, 0), 0)) {
      case (((n, s, e, w), max), step) => 
        val pos = map(step)((n, s, e, w))
        (pos, Math.max(max, getDistance(pos)))
    }

    println(s"part1 ${getDistance(pos)}")
    println(s"part2 $max")
  }
}

File: 2017/js/day-11.js
const fs = require('fs');
const input = fs.readFileSync('2017/input/day-10.input').toString();

const steps = input.split(',');
let map = {
  'n': ([n, s, e, w]) => ([n + 2, s - 2, e, w]),
  'ne': ([n, s, e, w]) => ([n + 1, s - 1, e + 1, w - 1]),
  'se': ([n, s, e, w]) => ([n - 1, s + 1, e + 1, w - 1]),
  's': ([n, s, e, w]) => ([n - 2, s + 2, e, w]),
  'sw': ([n, s, e, w]) => ([n - 1, s + 1, e - 1, w + 1]),
  'nw': ([n, s, e, w]) => ([n + 1, s - 1, e - 1, w + 1])
};
let pos = [0, 0, 0, 0];

const getDistance = pos => {
  const n = Math.abs(pos[0]);
  const w = Math.abs(pos[2]);
  return w + (n - w) / 2;
};
let max = 0;
steps.forEach(step => {
  pos = map[step](pos);
  max = Math.max(max, getDistance(pos));
});

console.log('part1', getDistance(pos));
console.log('part2', max);

